text/2020-03-07-copr-chunk.md

```diff
@@ -0,0 +1,150 @@
+2. SIMD instructions require continuous data in memory.
+3. Keep the same format with TiDB, so that it's possible to unify the computing framework between TiDB and TiKV.
+
+## Detailed Design
+
+Currently, An expression receives `&[RpnStackNode<'_>],` as input args, the definition of `RpnStackNode` is:
+
+```rust
+#[derive(Debug)]
+pub enum RpnStackNode<'a> {
+    /// Represents a scalar value. Comes from a constant node in expression list.
+    Scalar {
+        value: &'a ScalarValue,
+        field_type: &'a FieldType,
+    },
+
+    /// Represents a vector value. Comes from a column reference or evaluated result.
+    Vector {
+        value: RpnStackNodeVectorValue<'a>,
+        field_type: &'a FieldType,
+    },
+}
+```
+
+For non constant value, a column of data is represented by `RpnStackNodeVectorValue`:
+
+```rust
+#[derive(Debug)]
+pub enum RpnStackNodeVectorValue<'a> {
+    Generated {
+        // TODO: Maybe box it can be faster.
+        physical_value: VectorValue,
+    },
+    Ref {
+        physical_value: &'a VectorValue,
+        logical_rows: &'a [usize],
+    },
+}
+```
+
+Currently, the definition of `VectorValue` is:
+
+```rust
+#[derive(Debug, PartialEq, Clone)]
+pub enum VectorValue {
+    Int(Vec<Option<Int>>),
+    Real(Vec<Option<Real>>),
+    Decimal(Vec<Option<Decimal>>),
+    // TODO: We need to improve its performance, i.e. store strings in adjacent memory places
+    Bytes(Vec<Option<Bytes>>),
+    DateTime(Vec<Option<DateTime>>),
+    Duration(Vec<Option<Duration>>),
+    Json(Vec<Option<Json>>),
+}
+```
+
+We should modify the `VectorValue` to:
+
+```rust
+#[derive(Debug, PartialEq, Clone)]
+pub enum VectorValue {
+    Int(TypedVector<Int>),
+    Real(TypedVector<Real>),
+    Decimal(TypedVector<Decimal>),
+    Bytes(TypedVector<Bytes>),
+    DateTime(TypedVector<DateTime>),
+    Duration(TypedVector<Duration>),
+    Json(TypedVector<Json>),
+}
+```
+
+`TypedVector` implemented [chunk column format](https://pingcap.com/blog-cn/tidb-source-code-reading-10/).
+
+Although we can keep most API compatibility on `VectorValue` except `ScalarRef`. We have to modify the signature from
+
+```rust
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum ScalarValueRef<'a> {
+    Int(&'a Option<super::Int>),
+    // ...
+}
+```
+
+to
+
+```rust
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum ScalarValueRef<'a> {
+    Int(Option<&'a super::Int>),
+    // ...
+}
+```
+
+We also need to modify all rpn_function signature from
+
+```rust
+#[rpn_fn]
+pub fn md5(arg: &Option<Bytes>) -> Result<Option<Bytes>> {
+    // ...
+}
+```
+
+to
+
+```rust
+#[rpn_fn]
+pub fn md5(arg: Option<&Bytes>) -> Result<Option<Bytes>> {
```

Comment:
`args: &[u8]`

Reply:
This signature might not be very efficient for returning values in Chunk format, since in order to put these return values into a chunk column, it must be copied again.